### PR TITLE
CUDA parallel_reduce: Introduce volatile_wrapper to avoid need for join(volatile) overloads

### PR DESCRIPTION
**Pull request by PhilMiller**

### CODE REVIEW

**Comment by PhilMiller** [comment]
That doesn't square with my understanding of the CUDA docs that describe the effect of `volatile` as specifically requiring the compiler to emit cache-coherent load operations. I'll give it a try, though.

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 1]
```diff
@@ -854,6 +854,33 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
+      p_ret[i] = vpt[i];
+    }
+    return ret;
+  }
+
+  operator T() const { return get(); }
+
+  __device__ __host__ inline void set(const T& s) {
+    auto vps = reinterpret_cast<const volatile char*>(&s);
+    auto pt  = reinterpret_cast<char*>(&t);
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:877]:
```suggestion
    auto pt  = reinterpret_cast<volatile char*>(&t);
```

**Comment by PhilMiller** [comment]
@crtrott Even with the writes in `set()` going through a `volatile` pointer, the error persists.

**Comment by PhilMiller** [comment]
```
4: [ RUN      ] cuda.reducers_half_t
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:324: Failure
4: Expected equality of these values:
4:   sum_scalar
4:     Which is: 8.062500
4:   reference_sum
4:     Which is: 8.000000
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:324: Failure
4: Expected equality of these values:
4:   sum_scalar
4:     Which is: 487.500000
4:   reference_sum
4:     Which is: 498.000000
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:324: Failure
4: Expected equality of these values:
4:   sum_scalar
4:     Which is: 875.500000
4:   reference_sum
4:     Which is: 893.000000
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:324: Failure
4: Expected equality of these values:
4:   sum_scalar
4:     Which is: 1347.000000
4:   reference_sum
4:     Which is: 1379.000000
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:410: Failure
4: Expected equality of these values:
4:   prod_scalar
4:     Which is: 36.375000
4:   reference_prod
4:     Which is: 36.000000
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestReducers.hpp:410: Failure
4: Expected equality of these values:
4:   prod_scalar
4:     Which is: 15912.000000
4:   reference_prod
4:     Which is: 15552.000000
4: [  FAILED  ] cuda.reducers_half_t (3 ms)
4: [ RUN      ] cuda.reducers_int8_t
4: [       OK ] cuda.reducers_int8_t (3 ms)
4: [ RUN      ] cuda.reducers_point_t
4: [       OK ] cuda.reducers_point_t (2 ms)
4: [----------] 123 tests from cuda (43347 ms total)
4: 
4: [----------] Global test environment tear-down
4: [==========] 124 tests from 2 test suites ran. (43882 ms total)
4: [  PASSED  ] 123 tests.
4: [  FAILED  ] 1 test, listed below:
4: [  FAILED  ] cuda.reducers_half_t
4: 
4:  1 FAILED TEST
 4/54 Test  #4: KokkosCore_UnitTest_Cuda1 ....................***Failed  Error regular expression found in output. Regex=[  FAILED  ] 44.35 sec
```

**Comment by e10harvey** [comment]
@crtrott, @PhilMiller: With enough debugging prints uncommented in https://github.com/e10harvey/kokkos/tree/4077-volatile-reduce-type-wrapper, `cuda.reducers_half_t` passes.

@PhilMiller, just to make sure I'm understanding these changes correctly, the code is type-agnostic; correct? In this way, if we pass in an array of size `N` and type `T` where `T` is 2-bytes, (assuming the memory always responds in the same amount of time and the threads always get scheduled in the same fashion) then only operation affecting timing in the reduction is how long it takes to perform the arithmetic on `T` in each `join` operation; is that correct?

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 2]
```diff
@@ -854,6 +854,33 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
 namespace Kokkos {
 namespace Impl {
 
+template <typename T>
+struct alignas(T) volatile_wrapper {
+  T t;
+
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:860]: Consider adding copy ctor and operator=, or see what doesn't compile when they're `=delete;`

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp [hunk 3]
```diff
@@ -854,6 +854,33 @@ class ParallelFor<FunctorType, Kokkos::TeamPolicy<Properties...>,
...
+  __device__ __host__ inline volatile_wrapper(const T& rhs) { set(rhs.get()); }
+
+  __device__ __host__ inline T get() const {
+    T ret;
+    auto vpt   = reinterpret_cast<const volatile char*>(&t);
```
Comment (PhilMiller) [inline core/src/Cuda/Kokkos_Cuda_Parallel.hpp:865]: Try `const volatile T*`, etc, and skip the loop

File: core/src/Cuda/Kokkos_Cuda_Parallel.hpp (hunk 2 above)
Comment (PhilMiller) [reply core/src/Cuda/Kokkos_Cuda_Parallel.hpp:860]: Found at least one site that was doing direct assignment, so I implemented them.

**Comment by PhilMiller** [comment]
As of "Introduce volatile_wrapper usage for some of the lowest-level join calls", tests all pass, and one of the `join` call sites will no longer need to take a `volatile` argument

**Comment by PhilMiller** [comment]
Ok, in an optimized build, it turns out that the `half_t` reducer test *still* fails. Other than that, we're still looking good.

When I change another batch of code over to the wrappres, I get an additional failure in `cuda.mdrange_array_reduce` which is not too surprising that it would crop up. I think it may be an issue of trying to copy out a temporary of an array, and only capturing one element. That change is not yet pushed.

**Comment by PhilMiller** [comment]
Obviated by #4901

**Pull request closed without merging**